fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScanComparisons.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * IndexScanComparisons.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.IndexScanType;
+import com.apple.foundationdb.record.TupleRange;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.query.plan.ScanComparisons;
+import com.apple.foundationdb.record.query.plan.temp.explain.Attribute;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ScanComparisons} for use in an index scan.
+ */
+@API(API.Status.MAINTAINED)
+public class IndexScanComparisons implements IndexScanParameters {
+    @Nonnull
+    private final IndexScanType scanType;
+    @Nonnull
+    private final ScanComparisons scanComparisons;
+
+    public IndexScanComparisons(@Nonnull IndexScanType scanType, @Nonnull ScanComparisons scanComparisons) {
+        this.scanType = scanType;
+        this.scanComparisons = scanComparisons;
+    }
+
+    @Nonnull
+    public static IndexScanComparisons byValue() {
+        return byValue(null);
+    }
+
+    @Nonnull
+    public static IndexScanComparisons byValue(@Nullable ScanComparisons scanComparisons) {
+        if (scanComparisons == null) {
+            scanComparisons = ScanComparisons.EMPTY;
+        }
+        return new IndexScanComparisons(IndexScanType.BY_VALUE, scanComparisons);
+    }
+
+    @Nonnull
+    @Override
+    public IndexScanType getScanType() {
+        return scanType;
+    }
+
+    @Nonnull
+    public ScanComparisons getComparisons() {
+        return scanComparisons;
+    }
+
+    @Nonnull
+    @Override
+    public IndexScanRange bind(@Nonnull FDBRecordStoreBase<?> store, @Nonnull EvaluationContext context) {
+        return new IndexScanRange(scanType, scanComparisons.toTupleRange(store, context));
+    }
+
+    @Override
+    public int planHash(@Nonnull PlanHashKind hashKind) {
+        return scanType.planHash(hashKind) + scanComparisons.planHash(hashKind);
```

Comment:
This feels like the hashes should be combined with something like:

```java
PlanHashable.planHash(hashKind, scanType, scanComparisons);
```

Unless this is needed for plan hash stability purposes

Reply:
These are needed for stability.
```
-                return indexName.hashCode() + scanType.planHash(hashKind) + comparisons.planHash(hashKind) + (reverse ? 1 : 0);
+                return indexName.hashCode() + scanParameters.planHash(hashKind) + (reverse ? 1 : 0);
```